[docs] add v1.11 description to CHANGELOG.md (#5932)
DevExpress/testcafe: CHANGELOG.md

# Changelog

## v1.11.0 (03.02.2021)

### Enhancements

> #### :gear: Control Request Timeouts ([PR #5692](https://github.com/DevExpress/testcafe/pull/5692))

Reviewer: Can you actually "control" a value? You can control a feature or behavior. Indeed, you can "control" a value if you mean "persistently check and ensure it's okay". 

But meaning "specify" or "set"... I'm not sure

Author: `Set` is good

New CLI options allow you to control the timeout for requests. If a request is not met with a response within the specified period, TestCafe throws an error. The CLI options are:

* [--ajax-request-timeout](./docs/articles/documentation/reference/command-line-interface.md#--ajax-request-timeout-ms) controls the delay for fetch/XHR requests timeout
* [--page-request-timeout](./docs/articles/documentation/reference/command-line-interface.md#--page-request-timeout-ms) sets the timeout for receiving webpages.

Read the documentation for the corresponding configuration properties:

* [ajaxRequestTimeout](./docs/articles/documentation/reference/configuration-file.md#ajaxrequesttimeout)
* [pageRequestTimeout](./docs/articles/documentation/reference/configuration-file.md#ajaxrequesttimeout).

> The options are available in the [runner.run Method](./docs/articles/documentation/reference/testcafe-api/runner/run.md).

Reviewer: This whole piece is a little weird.

First, you begin by saying that `CLI` options are introduced (I wouldn't do that, because we add TestCafe options to all the interfaces).

Then, you introduce the information about config properties by saying "Read documentation...". I first thought it means: "To understand more about the aforementioned CLI options, read the documentation about the corresponding config properties".

I understand you don't want to write similar introductory phrases, but here you can go with just headings:

```
*CLI*

* options

example (!)

*Configuration file*

* options

example (!)

*JavaScript API*

* options

example (!)
```

Yes, I think it's better to add examples as well.

Reviewer 2: `The options are available in the [runner.run Method](./docs/articles/documentation/reference/testcafe-api/runner/run.md).`

What options are you describing? For example, are you referring to options that are listed below?

Author: The same options that are described a couple of lines above (CLI and Configuration File options)

Reviewer 2: Thanks for the clarification. Can we update it to one of the following?

`The above options are available in`...

or

`These options are available in`...

> #### :gear: Control Browser Initialization Timeout ([PR #5720](https://github.com/DevExpress/testcafe/pull/5720))

Reviewer: The same note about controlling a value

This release introduces an option to control browser initialization timeout. This timeout controls the time browsers have to connect to TestCafe before an error is thrown. You can control this timeout in one of the following ways:

* [--browser-init-timeout](./docs/articles/documentation/reference/command-line-interface.md#--browser-init-timeout-ms) CLI option

```sh
testcafe chrome my-tests --browser-init-timeout 180000
```

* [browserInitTimeout](./docs/articles/documentation/reference/configuration-file.md#browserinittimeout) configuration option

```json
{
    "browserInitTimeout": 180000
}
```

* [runner.run Method](./docs/articles/documentation/reference/testcafe-api/runner/run.md) parameter

```js
runner.run({ "browserInitTimeout": 180000 })
```

This setting sets an equal timeout for local and [remote browsers](./docs/articles/documentation/guides/concepts/browsers.md#browsers-on-remote-devices).

> #### Improved the `Unable To Establish Browser Connection` Error Message ([PR #5720](https://github.com/DevExpress/testcafe/pull/5720))

Reviewer: We usually don't include error message enhancements in release notes considering them too minor. But maybe it's a good idea to finally do it if the devs don't mind.

Reviewer 2: ```suggestion
#### Improve the `Unable To Establish Browser Connection` Error Message ([PR #5720](https://github.com/DevExpress/testcafe/pull/5720))
```

TestCafe raises this error when at least one local or remote browser wasn't able to connect. The error message now includes the number of browsers that have not established a connection.

TestCafe raises a warning if low system performance is causing the connectivity issue.

#### :gear: An Option to Retry Requests for the Test Page ([PR #5738](https://github.com/DevExpress/testcafe/pull/5738))

> If the first request for the tested webpage is not satisfied, TestCafe can now retry this request.

Reviewer: I think it's simply "responded", not "satisfied".

I would say "a request is satisfied" if I meant "The response has been received and the sender is satisfied with the response"... 😄

Author: If a tested webpage wasn't served after the first request, TestCafe can now retry this request.

# Changelog

## v1.11.0 (03.02.2021)

### Enhancements

#### :gear: Set Request Timeouts ([PR #5692](https://github.com/DevExpress/testcafe/pull/5692))

> TestCafe now enables you to set request timeouts. If a request is not met with a response within the specified period, TestCafe throws an error.

Reviewer 2: What do you mean by "not met with a response"? Do you mean that the request does not receive a response? For example, can we say:

`If a request does not receive a response within the specified period, TestCafe throws an error.`

Author: Is it correct to say that a request receives a response? Or is it more correct to say `the issuer of the request receives a response`? In that case, can we say:

> If TestCafe receives no response within the specified period, it throws an error.

Reviewer 2: > If TestCafe receives no response within the specified period, it throws an error.

I like your suggestion. Please apply it. Thanks.

*CLI*

* [--ajax-request-timeout](./docs/articles/documentation/reference/command-line-interface.md#--ajax-request-timeout-ms) controls the timeout for fetch/XHR requests

> * [--page-request-timeout](./docs/articles/documentation/reference/command-line-interface.md#--page-request-timeout-ms) sets the timeout for webpages requests

Reviewer 2: ```suggestion
* [--page-request-timeout](./docs/articles/documentation/reference/command-line-interface.md#--page-request-timeout-ms) sets the timeout for webpage requests
```

}
```

#### :gear: Set Browser Initialization Timeout ([PR #5720](https://github.com/DevExpress/testcafe/pull/5720))

This release introduces an option to control browser initialization timeout. This timeout controls the time browsers have to connect to TestCafe before an error is thrown. You can control this timeout in one of the following ways:

* [--browser-init-timeout](./docs/articles/documentation/reference/command-line-interface.md#--browser-init-timeout-ms) CLI option

```sh
testcafe chrome my-tests --browser-init-timeout 180000
```

* [browserInitTimeout](./docs/articles/documentation/reference/configuration-file.md#browserinittimeout) configuration option

```json
{
    "browserInitTimeout": 180000
}
```

* [runner.run Method](./docs/articles/documentation/reference/testcafe-api/runner/run.md) parameter

```js
runner.run({ "browserInitTimeout": 180000 })
```

This setting sets an equal timeout for local and [remote browsers](./docs/articles/documentation/guides/concepts/browsers.md#browsers-on-remote-devices).

#### Improved the `Unable To Establish Browser Connection` Error Message ([PR #5720](https://github.com/DevExpress/testcafe/pull/5720))

> TestCafe raises this error when at least one local or remote browser wasn't able to connect. The error message now includes the number of browsers that have not established a connection.

Reviewer 2: ```suggestion
TestCafe raises this error when at least one local or remote browser was not able to connect. The error message now includes the number of browsers that have not established a connection.
```

TestCafe raises a warning if low system performance is causing the connectivity issue.

#### :gear: An Option to Retry Requests for the Test Page ([PR #5738](https://github.com/DevExpress/testcafe/pull/5738))

> If a tested webpage wasn't served after the first request, TestCafe can now retry the request.

Reviewer 2: ```suggestion
If a tested webpage was not served after the first request, TestCafe can now retry the request.
```

    ```sh
    testcafe chrome test.js --retry-test-pages
    ```

* the [runner.run](https://devexpress.github.io/testcafe/documentation/using-testcafe/programming-interface/runner.html) option

    ```js
    runner.run({
        retryTestPages: true
    });
    ```

* the [retryTestPages](https://devexpress.github.io/testcafe/documentation/using-testcafe/configuration-file.html#retrytestpages) configuration file property

    ```json
    {
        "retryTestPages": true
    }
    ```

### Bug Fixes

* Fixed a bug where `Selector.withText` couldn't locate elements inside an `iframe` ([#5886](https://github.com/DevExpress/testcafe/issues/5886))
* Fixed a bug where TestCafe was sometimes unable to detect when a browser instance closes ([#5857](https://github.com/DevExpress/testcafe/issues/5857))
* You can now install TestCafe with `Yarn 2` ([PR #5872](https://github.com/DevExpress/testcafe/pull/5872) by [@NiavlysB](https://github.com/NiavlysB))
* Fixed a bug where TestCafe was sometimes unable to create a `Web Worker` from an object ([testcafe-hammerhead/#2512](https://github.com/DevExpress/testcafe-hammerhead/issues/2512))

> * Fixed an error thrown by TestCafe proxy when trying to delete an object property that doesn't exist ([testcafe-hammerhead/#2504](https://github.com/DevExpress/testcafe-hammerhead/issues/2504))

Reviewer 2: ```suggestion
* Fixed an error thrown by TestCafe proxy when trying to delete an object property that does not exist ([testcafe-hammerhead/#2504](https://github.com/DevExpress/testcafe-hammerhead/issues/2504))
```

> * Fixed an error thrown  by TestCafe proxy when a Service Worker overwrites properties of a `window` object ([testcafe-hammerhead/#2538](https://github.com/DevExpress/testcafe-hammerhead/issues/2538))

Reviewer 2: ```suggestion
* Fixed an error thrown by TestCafe proxy when a Service Worker overwrites properties of a `window` object ([testcafe-hammerhead/#2538](https://github.com/DevExpress/testcafe-hammerhead/issues/2538))
```

* Fixed a bug where `t.openWindow` method requested a URL twice ([testcafe-hammerhead/#2544](https://github.com/DevExpress/testcafe-hammerhead/issues/2544))
* Fixed an error (`TypeError: Illegal invocation`) thrown by TestCafe on pages that contain an XMLDocument with an `iframe` ([testcafe-hammerhead/#2554](https://github.com/DevExpress/testcafe-hammerhead/issues/2554))
* Fixed an error (`SyntaxError: Identifier has already been declared`) thrown by TestCafe on pages with scripts that create nested JavaScript objects ([testcafe-hammerhead/#2506](https://github.com/DevExpress/testcafe-hammerhead/issues/2506))
* Fixed a bug where TestCafe was unable to focus elements within shadow DOM ([testcafe-hammerhead/#2408](https://github.com/DevExpress/testcafe-hammerhead/issues/2408))

> * TestCafe throws an error when an entity of type other than `Error` is thrown in a test script ([PR testcafe-hammerhead/#2536](https://github.com/DevExpress/testcafe-hammerhead/pull/2536))

Reviewer 2: Are you reporting a bug fix, or just stating an error that is thrown?

Author: This is an improvement. TestCafe used not to handle this case (a generic error was thrown, and it was quite hard to determine the cause). Now the error is explicit.

Does this work:

> TestCafe **now** throws an error when an entity of type other than `Error` is thrown in a test script ([PR testcafe-hammerhead/#2536](https://github.com/DevExpress/testcafe-hammerhead/pull/2536))

Reviewer 2: > TestCafe now throws an error when an entity of type other than Error is thrown in a test script (PR testcafe-hammerhead/#2536)

Thanks for the clarification. Please apply your suggestion.

* TestCafe proxy now correctly handles requests that are not permitted by the CORS policy ([testcafe-hammerhead/#1263](https://github.com/DevExpress/testcafe-hammerhead/issues/1263))
* Improved compatibility with test pages that use `with` statements ([testcafe-hammerhead/#2434](https://github.com/DevExpress/testcafe-hammerhead/issues/2434))
* TestCafe proxy can now properly parse statements that use a comma operator in `for..of` loops ([testcafe-hammerhead/#2573](https://github.com/DevExpress/testcafe-hammerhead/issues/2573))
* Fixed a bug where TestCafe would open a new window even if `preventDefault` is present in element's event handler ([testcafe-hammerhead/#2582](https://github.com/DevExpress/testcafe-hammerhead/pull/2582))  

### Vulnerability Fix ([PR #5843](https://github.com/DevExpress/testcafe/pull/5843), [PR testcafe-hammerhead#2531](https://github.com/DevExpress/testcafe-hammerhead/pull/2531))

We have fixed a vulnerability found in the [debug](https://www.npmjs.com/package/debug) module we use for debugging.

> The vulnerability was a [ReDos Vulnerability Regression](https://github.com/visionmedia/debug/issues/797) and affected all TestCafe users. TestCafe now uses `debug@4.3.1`, where the issue is fixed.

Reviewer 2: ```suggestion
The vulnerability was a [ReDos Vulnerability Regression](https://github.com/visionmedia/debug/issues/797) that affected all TestCafe users. TestCafe now uses `debug@4.3.1`, where the issue is fixed.
```

* the [--retry-test-pages](https://devexpress.github.io/testcafe/documentation/using-testcafe/command-line-interface.html#--retry-test-pages) command line argument

    ```sh
    testcafe chrome test.js --retry-test-pages
    ```

* the [runner.run](https://devexpress.github.io/testcafe/documentation/using-testcafe/programming-interface/runner.html) option

    ```js
    runner.run({
        retryTestPages: true
    });
    ```

* the [retryTestPages](https://devexpress.github.io/testcafe/documentation/using-testcafe/configuration-file.html#retrytestpages) configuration file property

    ```json
    {
        "retryTestPages": true
    }
    ```

### Bug Fixes

* Fixed a bug where `Selector.withText` couldn't locate elements inside an `iframe` ([#5886](https://github.com/DevExpress/testcafe/issues/5886))
* Fixed a bug where TestCafe was sometimes unable to detect when a browser instance closes ([#5857](https://github.com/DevExpress/testcafe/issues/5857))
* You can now install TestCafe with `Yarn 2` ([PR #5872](https://github.com/DevExpress/testcafe/pull/5872) by [@NiavlysB](https://github.com/NiavlysB))

> * Fixed a bug where the `typeText` action would sometimes not be able to replace existing text ([PR #5942](https://github.com/DevExpress/testcafe/pull/5942) by [@rueyaa332266](https://github.com/rueyaa332266))

Author: @albertov05  A new entry in this PR since your last visit, please take a look

Reviewer 2: I'd like to simplify "would sometimes not be able to replace", if possible. What do you think about the following?

`Fixed a bug where the `typeText` action does not always replace existing text`

Author: awesome

Reviewer 2: Thank you.

* The `window.history.replaceState` function is overridden incorrectly ([testcafe-hammerhead/#1146](https://github.com/DevExpress/testcafe-hammerhead/issues/1146))
* Hammerhead crashes if a script file contains a sourcemap comment ([testcafe-hammerhead/#1052](https://github.com/DevExpress/testcafe-hammerhead/issues/1052))
* The proxy should override the `DOMParser.parseFromString` method ([testcafe-hammerhead/#1133](https://github.com/DevExpress/testcafe-hammerhead/issues/1133))

> * The `fetch` method should emulate the native behavior on merging headers ([testcafe-hammerhead/#1116](https://github.com/DevExpress/testcafe-hammerhead/issues/1116))

Reviewer 2: To avoid the -ing form ("merging"), can we change "merging headers" to "header merge"?

> * The `fetch` method should emulate the native behavior on merging headers ([testcafe-hammerhead/#1116](https://github.com/DevExpress/testcafe-hammerhead/issues/1116))

Author: ```suggestion
* The `fetch` method should emulate the native behavior on header merge ([testcafe-hammerhead/#1116](https://github.com/DevExpress/testcafe-hammerhead/issues/1116))
```
